### PR TITLE
Allow using interface-converted type-requirement in fields

### PR DESCRIPTION
**Pull request by SupunS**
Closes https://github.com/onflow/cadence/issues/3527

(Solves https://flow-foundation.slack.com/archives/C074F78ECNL/p1723566139471899)

If a type-requirement has been converted to an interface, then allow using such updated type in field. i.e: allow changing the field types to the interface type. Here only `T` would become `{T}`, and nothing else would be allowed.

### CODE REVIEW

File: cmd/util/ledger/migrations/cadence.go [hunk 1]
```diff
@@ -41,6 +41,28 @@ func NewInterfaceTypeConversionRules(chainID flow.ChainID) StaticTypeMigrationRu
 	}
 }
 
+var typeRequirementRemovals = []struct {
```
Comment (turbolent) [inline cmd/util/ledger/migrations/cadence.go:44]: Nice! Maybe we can generalize this and infer it from an update (analyze old -> new)? 

Basically all nested composite types that were declared inside of a contract interface can be assumed to have to be changed to interface types.

**Review by turbolent — COMMENTED** [review]
Nice!

File: cmd/util/ledger/migrations/cadence.go (hunk 1 above)
Comment (SupunS) [reply cmd/util/ledger/migrations/cadence.go:44]: yeah, can do that. Going to need another migration step (doesn't migrate anything), to extract this from the existing payloads. That needs to be run before the staged contract migration.

Comment (SupunS) [reply cmd/util/ledger/migrations/cadence.go]: added in the recent commits

**Comment by SupunS** [comment]
Need to add some more tests

File: cmd/util/ledger/migrations/static_type_migration.go [hunk 2]
```diff
@@ -8,18 +8,22 @@ import (
 type StaticTypeMigrationRules map[common.TypeID]interpreter.StaticType
 
 func NewStaticTypeMigration[T interpreter.StaticType](
-	rules StaticTypeMigrationRules,
+	rulesGetter func() StaticTypeMigrationRules,
```
Comment (turbolent) [inline cmd/util/ledger/migrations/static_type_migration.go:11]: Is there a particular reason why this is a getter, and the rules are not passed as-is?

File: cmd/util/ledger/migrations/type_requirements_extractor.go [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+	typeRequirements []TypeRequirement
+}
+
+type TypeRequirement struct {
+	Address      common.Address
+	ContractName string
+	TypeName     string
+}
+
+func NewTypeRequirementsExtractingMigration(
```
Comment (turbolent) [inline cmd/util/ledger/migrations/type_requirements_extractor.go:32]: Do we need a new migration for this? Could the extraction be integrated into the existing checking migration?

**Review by turbolent — COMMENTED** [review]
Great work generalizing this! 👏

**Review by janezpodhostnik — COMMENTED** [review]
Looks good to me, but I'm also wondering if we can join the  `TypeRequirementsExtractingMigration` with an existing migration

File: cmd/util/ledger/migrations/type_requirements_extractor.go (hunk 3 above)
Comment (SupunS) [reply cmd/util/ledger/migrations/type_requirements_extractor.go:32]: I initially had the same thought, but the existing checking migration runs for the new code (i.e: after the staged-contracts migration), but we need to extract the type-requriements from the old code (i.e: before the staged contracts migration)

File: cmd/util/ledger/migrations/static_type_migration.go (hunk 2 above)
Comment (SupunS) [reply cmd/util/ledger/migrations/static_type_migration.go:11]: This constructor is called at the time of the constructing the migrations (i.e: Cadence value migration). But the rules can only be finalized after running this newly added migration. i.e: the passed `LegacyTypeRequirements` list is lazily loaded. So we need to delay the construction of the rules, until after the execution of previous 
 migration (`TypeRequirementsExtractingMigration` to be exact).

File: cmd/util/ledger/migrations/type_requirements_extractor.go (hunk 3 above)
Comment (SupunS) [reply cmd/util/ledger/migrations/type_requirements_extractor.go:32]: cc: @janezpodhostnik

**Review by janezpodhostnik — APPROVED** [review]

File: cmd/util/ledger/migrations/type_requirements_extractor.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+		sort.Slice(contracts, func(i, j int) bool {
+			a := contracts[i]
+			b := contracts[j]
+			return a.location.ID() < b.location.ID()
+		})
+
+		log.Info().Msgf("Gathered all contracts (%d)", len(contracts))
+
+		// Check all contracts
```
Comment (turbolent) [inline cmd/util/ledger/migrations/type_requirements_extractor.go:110]:
```suggestion
		// Extract type requirements from all contracts
```

File: cmd/util/ledger/migrations/static_type_migration.go (hunk 2 above)
Comment (turbolent) [reply cmd/util/ledger/migrations/static_type_migration.go:11]: I see! Can you please leave some comments like this explanation in the code?

File: cmd/util/ledger/migrations/type_requirements_extractor.go (hunk 3 above)
Comment (turbolent) [reply cmd/util/ledger/migrations/type_requirements_extractor.go:32]: Ah, I see now that this new migration only gathers the contract codes from all registers and extracts the type requirements syntactically. In that case we need a new migration, right. 

Could we maybe still refactor out the common part, the extraction of contract codes from all registers?

**Review by turbolent — COMMENTED** [review]
Great work!

File: cmd/util/ledger/migrations/type_requirements_extractor.go (hunk 3 above)
Comment (SupunS) [reply cmd/util/ledger/migrations/type_requirements_extractor.go:26]: good idea! added in https://github.com/onflow/flow-go/pull/6341/commits/6736db422e5e4aa2d43e4fb019e06c74217ade64

**Review by turbolent — APPROVED** [review]
Great work!

**Pull request merged**